### PR TITLE
Integrate leak_tracker.

### PR DESCRIPTION
**Pull request by polina-c**
See UX in description of https://github.com/dart-lang/leak_tracker/pull/20

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart [hunk 1]
```diff
@@ -6,78 +6,81 @@ import 'dart:async';
...
 
-  void _receivedLeaksSummary(Event event) {
-    if (event.extensionKind != _extensionKindToReceiveLeaksSummary) return;
-    leakSummaryReceived.value = true;
-    try {
-      final newSummary = LeakSummary.fromJson(event.json!['extensionData']!);
-      final time = event.timestamp != null
-          ? DateTime.fromMicrosecondsSinceEpoch(event.timestamp!)
-          : DateTime.now();
+  static DateTime _fromEpochOrNull(int? microsecondsSinceEpoch) =>
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart:53]: Nit: maybe `_fromEpochOrNow`? `_fromEpochOrNull` is a similar name to `firstWhereOrNull` which returns null, whereas `_fromEpochOrNull` doesn't have a nullable return type.

File: packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart [hunk 2]
```diff
@@ -6,78 +6,81 @@ import 'dart:async';
...
+      appProtocolVersion = message.protocolVersion;
+      return;
+    }
+
+    if (message is LeakSummary) {
+      appStatus.value = AppStatus.leaksFound;
+      if (message.matches(_lastLeakSummary)) return;
+      _lastLeakSummary = message;
+
+      final time = _fromEpochOrNull(vmServiceEvent.timestamp).toLocal();
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart:75]: Have you actually encountered a `null` timestamp? The [service spec](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#event) states that this field is always provided for `Event`.

File: packages/devtools_app/lib/src/screens/memory/panes/leaks/primitives/simple_items.dart [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+enum AppStatus {
+  leakTrackingNotSupported,
+  noCommunicationsRecieved,
+  unsupportedProtocolVersion,
+  leakTrackingStarted,
+  leaksFound,
+}
+
+/// We may use https://pub.dev/packages/vendor to support previous versions of leak_tracker.
+const supportedLeakTrackingProtocols = {'1'};
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/memory/panes/leaks/primitives/simple_items.dart:14]: Nit: add the type argument to the set declaration

File: packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart (hunk 1 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart]: deleted this method

File: packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart [hunk 4]
```diff
@@ -6,78 +6,81 @@ import 'dart:async';
...
+          ? DateTime.now()
+          : DateTime.fromMicrosecondsSinceEpoch(microsecondsSinceEpoch);
+
+  void _onAppMessageWithHistory(Event vmServiceEvent) {
+    if (appStatus.value == AppStatus.unsupportedProtocolVersion) return;
+
+    final message = EventFromApp.fromVmServiceEvent(vmServiceEvent)?.message;
+    if (message == null) return;
+
+    if (message is LeakTrackingStarted) {
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart:59]: Will do this as patterns get released. Fir now there is no type promotion for 'runtimeType':

Screenshot 2022-12-07 at 9 51 25 AM

[the remark this answers is not in the document]
Comment (bkonyi) [reply packages/devtools_app/lib/src/screens/memory/panes/leaks/controller.dart:59]: Ah good point. Waiting for patterns SGTM.

**Review by CoderDake — APPROVED** [review]

**Pull request merged**